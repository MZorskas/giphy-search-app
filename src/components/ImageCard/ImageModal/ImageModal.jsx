import React, { useEffect, useState, useRef } from 'react';
import './ImageModal.scss';
import ReactDOM from 'react-dom';
import { CloseIcon } from 'assets/icons';
import IconButton from 'components/IconButton';
import PropTypes from 'prop-types';

const focusableNodes = ['button', 'a', 'input', 'select', 'textarea'];

const ImageModal = ({ imageUrl, closeModal, alt }) => {
  const modalRef = useRef(null);
  const closeButtonRef = useRef(null);

  const [lastFocusableNode, setLastFocusableNode] = useState();

  const onClickAway = (event) => {
    if (modalRef.current && !modalRef.current.contains(event.target))
      closeModal();
  };

  const onKeyDown = (e) => {
    // ESC button
    if (e.keyCode === 27) closeModal();
    // TAB + SHIFT / TAB buttons
    if (e.keyCode === 9 && e.shiftKey && e.target === lastFocusableNode[0]) {
      e.preventDefault();
      lastFocusableNode[lastFocusableNode.length - 1].focus();
    } else if (
      e.keyCode === 9 &&
      !e.shiftKey &&
      e.target === lastFocusableNode[lastFocusableNode.length - 1]
    ) {
      e.preventDefault();
      lastFocusableNode[0].focus();
    }
  };

  useEffect(() => {
    closeButtonRef.current.focus();
    if (modalRef)
      setLastFocusableNode(
        Array.from(modalRef.current.childNodes).filter((node) =>
          focusableNodes.includes(node.localName)
        )
      );
  }, [closeButtonRef, modalRef, setLastFocusableNode]);

  return ReactDOM.createPortal(
    <aside className="image-modal" onClick={onClickAway} onKeyDown={onKeyDown}>
      <div className="image-modal__content" ref={modalRef}>
        <IconButton
          className="image-modal__button"
          icon={<CloseIcon />}
          onClick={closeModal}
          ref={closeButtonRef}
        />
        <img src={imageUrl} alt={alt} className="image-modal__image" />
      </div>
    </aside>,
    document.getElementById('modal-root')
  );
};

ImageModal.propTypes = {
  alt: PropTypes.string,
  imageUrl: PropTypes.string.isRequired,
  closeModal: PropTypes.func.isRequired,
};

export default ImageModal;
