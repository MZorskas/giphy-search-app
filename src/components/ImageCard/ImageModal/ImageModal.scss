@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.image-modal {
  $_modal-spacing: 2rem;

  @include position-fixed(0, null, null, 0);

  background: rgba(0, 0, 0, 0.8);
  display: flex;
  height: 100%;
  width: 100%;
  z-index: 100;

  &__content {
    $_max-content-size: 60rem;

    @include position-fixed(50%, null, null, 50%);

    box-sizing: border-box;
    display: flex;
    flex-direction: column;
    max-width: $_max-content-size;
    padding: $_modal-spacing;
    transform: translate(-50%, -50%);
    width: 100%;
  }

  &__button {
    @include position-absolute(0, $_modal-spacing, null, null);

    padding: 0;
  }

  &__image {
    border: 1px solid $color-primary;
  }
}
