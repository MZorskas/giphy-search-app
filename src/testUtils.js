export const generateModalRoot = () => {
  const modalRoot = global.document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  const body = global.document.querySelector('body');
  body.appendChild(modalRoot);
};
