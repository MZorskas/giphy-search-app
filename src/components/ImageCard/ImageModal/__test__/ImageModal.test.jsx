import React from 'react';
import ReactDom from 'react-dom';
import ImageModal from '../ImageModal';
import { generateModalRoot } from 'testUtils';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
// import '@testing-library/jest-dom/extend-expect';

const imageUrl = 'imageUrl';
const alt = 'alt';

describe('ImageModal component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    generateModalRoot();
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
    render(<ImageModal />);
    expect(screen.queryByTestId('image-modal')).toBeInTheDocument();
    expect(spy).toHaveBeenCalled();
  });

  it('renders without propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    const mockOnClick = jest.fn();
    generateModalRoot();
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
    render(<ImageModal imageUrl={imageUrl} closeModal={mockOnClick} />);
    expect(screen.queryByTestId('image-modal')).toBeInTheDocument();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders and focuses IconButton', () => {
    const mockOnClick = jest.fn();
    generateModalRoot();
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
    render(
      <ImageModal imageUrl={imageUrl} alt={alt} closeModal={mockOnClick} />
    );
    expect(screen.queryByTestId('image-modal')).toBeInTheDocument();
    expect(document.activeElement).toEqual(screen.queryByTestId('icon-button'));
  });

  it('fires closeModal function prop with escape key', () => {
    const mockOnClick = jest.fn();
    generateModalRoot();
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();
    render(<ImageModal imageUrl={imageUrl} closeModal={mockOnClick} />);
    expect(mockOnClick).toHaveBeenCalledTimes(0);
    fireEvent.keyDown(screen.queryByTestId('image-modal'), {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    });
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('fires closeModal function prop when clicking outsite of modal', () => {
    const mockOnClick = jest.fn();
    generateModalRoot();
    render(
      <ImageModal imageUrl={imageUrl} closeModal={mockOnClick} alt={alt} />
    );
    expect(mockOnClick).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.queryByTestId('image-modal-content'));
    fireEvent.click(screen.getByAltText(alt));
    expect(mockOnClick).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.queryByTestId('image-modal'));
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('fires closeModal function prop when clicking on IconButton', () => {
    const mockOnClick = jest.fn();
    generateModalRoot();
    render(
      <ImageModal imageUrl={imageUrl} closeModal={mockOnClick} alt={alt} />
    );
    expect(mockOnClick).toHaveBeenCalledTimes(0);
    fireEvent.click(screen.queryByTestId('icon-button'));
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    ReactDom.createPortal = jest.fn((element, node) => {
      return element;
    });
    const tree = renderer
      .create(<ImageModal imageUrl={imageUrl} alt={alt} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
