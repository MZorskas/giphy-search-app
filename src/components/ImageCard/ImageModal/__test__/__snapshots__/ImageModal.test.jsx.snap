// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageModal component matches snapshot 1`] = `
<aside
  className="image-modal"
  data-testid="image-modal"
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div
    className="image-modal__content"
    data-testid="image-modal-content"
  >
    <button
      aria-label="search"
      className="icon-button image-modal__button"
      data-testid="icon-button"
      type="button"
    >
      <svg>
        symbol_x.svg
      </svg>
    </button>
    <img
      alt="alt"
      className="image-modal__image"
      src="imageUrl"
    />
  </div>
</aside>
`;
